File: singleton/atividade/singleton_joao_victor_negreiros.cpp
#include <iostream>
using std::cin;
using std::cout;
using std::endl;

class Highlander {
  private:
    Highlander() {};
    static Highlander* instancia;
    static int invocacoes;

  public:
    static Highlander* obterInstancia() {
      if (instancia == nullptr) {
        instancia = new Highlander();
      }

      invocacoes += 1;
      return instancia;
    }

    static int totalInvocacoes() {
      return invocacoes;
    }
};

// Inicializacao dos atributos de Highlander
Highlander* Highlander::instancia = nullptr;
int Highlander::invocacoes = 0;


class Fabrica {
  public:
    static void main() {
      int numInstancias;

      cout << "Qtd de instancias: ";
      cin >> numInstancias;

      for (int i = 0; i < numInstancias; i++) {
        Highlander::obterInstancia();
      }
      
      cout << "---------------------" << endl;
      cout << "Total de invocacoes: " << Highlander::totalInvocacoes() << endl;
      
    }
};

int main(){
  Fabrica::main();

  return 0;
}

File: singleton/atividade/singleton_joao_victor_negreiros.ts
class Highlander {
  private constructor () {}

  private static instancia: Highlander;

  private invocacoes = 0;

  static obterInstancia(): Highlander {
    if (!this.instancia) {
      this.instancia = new Highlander()
    }

    this.instancia.invocacoes += 1;
    return this.instancia
  }


  static totalInvocacoes(): number {
    return Highlander.instancia.invocacoes;
  }
}


class Fabrica {
  static main(num: number): void {
    for (let i = 0; i < num; i++) {
      Highlander.obterInstancia()
    }

    const total = Highlander.totalInvocacoes()

    console.log('Input: ', num);
    console.log('Total de invocacoes: ', total);
  }
}

// Main
Fabrica.main(5)